### PR TITLE
Aurea & Anibel - Edges

### PR DESCRIPTION
**Pull request by Aurea-Li**
# OO Ride Share
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
:------------- | :-------------
Describe a design decision you had to make when working on this project. What options were you considering? What helped you make your final decision? | For `duration` in the `Trip` class, we decided that for any in-progress trips, it would evaluate to `nil` as opposed to `0`. We used `nil` as an indication of in-progress trips in our other method `total_time_spent`. 
Describe any examples of composition that you encountered in this project, if any | Each `User` has a one to many relationship to `Trip`, and each `Trip` has a one to one relationship with a `Driver`, and a one to one or one to many relationship with `User` as a passenger.  
Describe the relationship between `User` and `Driver`| `Driver` inherits `User` attributes and behaviors. It is a direct child of the `User` class. 
Describe a _nominal test_ that you wrote for this assignment. | In `trip_dispatcher_spec.rb`, in the `assign_driver` method tests starting on line 177, we wrote tests that checked to make sure that it returns the driver with the oldest most recent trip. 
Describe an _edge case test_ that you wrote for this assignment | In `trip_dispatcher_spec.rb`, the test starting in line 136, we check to see that `request_trip` returns `nil` if no available drivers. 
Describe a concept that you/your pair gained more clarity on as you worked on this assignment | When we worked on wave 2 and updating/writing the `load` methods for the three classes, it became apparent to think about the order in which we loaded csv files since the classes interacted with each other. 
What are two discussion points that you and your pair discussed when giving/receiving feedback from each other that you would be willing to share? | One thing we struggled with was how to pair successfully when we are trying to debug nested and/or embedded objects/data structures. It was also mentioned that debugging could be daunting in front of an audience. We could also work on letting the driver try to solve a coding problem first before the navigator directs them.

### CODE REVIEW

**Comment by tildeee** [comment]
# Ride Share
## What We're Looking For

Feature|Feedback
:------------- | :-------------
**Baseline** |
Used Git Regularly | x
Answer comprehension questions | x
**Wave 1** |
Appropriate use of Ruby's `Date`| x
`Trip` has a helper method to calculate duration| x
`User` has a method to calculate total cost of all trips| x
Tests for wave 1 | x
**Wave 2** |
`Driver` inherits from `User` | x
`Driver` has `add_driven_trip` method | x
`Driver` has method to calculate average rating | x
`Driver` has method to calculate net expenditures and it uses `super` | x
`Driver` has a method to calculate total revenue | x
Tests for wave 2 | x
**Wave 3** |
`TripDispatcher` has a new method to create trips | x
creating a trip in `TripDispatcher` relies on methods in `Driver` and `User` (passenger) to modify their own attributes | 
Complex logic was correctly implemented | x
Tests for `request_trip` | x, well done!
Methods from wave 1 and 2 handle incomplete trips | x
Tests for wave 1 and 2 methods with incomplete trips | x
**Wave 4 (Optional)** |
`TripDispatcher` now assigns trips to either the newest driver (no trips), or the driver who has not driven in the longest time | x
Appropriate helper methods were made to help with complex logic | x
Tests for wave 4 | x
**Overall** | 

Great work on this project you too!

Your code is really clear and readable-- in the bits with complex logic, the code ends up being very clear because of the appropriate use of `Enumerable` short hand methods, and otherwise in general everything else was kept tidy and organized.

The ternaries really paid off in this code base-- everything looks great

You made all of your instance variables in every class read-only, and that makes me really happy!

Adding a few comments for suggestions, but otherwise good job

File: lib/driver.rb [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+    end
+
+    def change_status
+      @status == :AVAILABLE ? @status = :UNAVAILABLE : @status = :AVAILABLE
+    end
+
+    def total_revenue
+      valid_trips = @driven_trips.find_all { |trip| trip.cost }
+
+      return (valid_trips.sum { |trip| trip.cost - 1.65} * 0.80).round(2)
```
Comment (tildeee) [inline lib/driver.rb:48]: I'd probably pull these "magic number" integers out into constants

File: lib/trip_dispatcher.rb [hunk 2]
```diff
@@ -1,16 +1,18 @@
...
 module RideShare
   class TripDispatcher
     attr_reader :drivers, :passengers, :trips
 
     def initialize(user_file = 'support/users.csv',
-                   trip_file = 'support/trips.csv')
+                   trip_file = 'support/trips.csv',
+                   driver_file = 'support/drivers.csv')
       @passengers = load_users(user_file)
+      @drivers = load_drivers(driver_file)
```
Comment (tildeee) [inline lib/trip_dispatcher.rb:15]: Nicely done loading drivers before trips!

File: specs/trip_dispatch_spec.rb [hunk 3]
```diff
@@ -111,4 +115,104 @@
...
+      driver = @dispatcher.find_driver(8)
+      num_of_trips = driver.driven_trips.length
+      @dispatcher.request_trip(1)
+
+      expect( driver.driven_trips.length - num_of_trips ).must_equal 1
+    end
+
+    it 'selects a driver with an available status' do
+      driver = @dispatcher.find_driver(8)
+      num_of_trips = driver.driven_trips.length
```
Comment (tildeee) [inline specs/trip_dispatch_spec.rb:164]: Copy/pasta! `num_of_trips` doesn't get used in this test

File: specs/trip_dispatch_spec.rb [hunk 4]
```diff
@@ -111,4 +115,104 @@
...
+      expect( @dispatcher.assign_driver.id ).must_equal 5
+
+      trip3 = RideShare::Trip.new(id: 8, driver: driver_5, passenger: nil,
+                                  start_time: Time.parse("2018-05-25 11:00:00 -0700"),
+                                  end_time: Time.parse("2018-05-25 11:30:02 -0700"),
+                                  rating: 1)
+
+      driver_5.add_driven_trip(trip3)
+
+      expect( @dispatcher.assign_driver.id ).must_equal 8
```
Comment (tildeee) [inline specs/trip_dispatch_spec.rb:215]: Great job with these tests and making the test data yours so you can use concrete answers
